### PR TITLE
Make Slow Requests an extension point.

### PR DESCRIPTION
**Pull request by christ66**
I would like `SlowRequests` to be an extension point, so we can add other checks in the event a slow request happens. We can add checks like I/O, networking, etc. when there is a slow request.

@reviewbybees

### CODE REVIEW

**Comment by JTSweet** [comment]
🐝

**Comment by jglick** [comment]
So where is the downstream PR showing this being used?

**Comment by christ66** [comment]
@jglick nothing yet. Looking to implement it as an extension point first, then we'd need to figure out what needs to be executed during slow requests.

File: src/main/java/com/cloudbees/jenkins/support/slowrequest/SlowRequest.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package com.cloudbees.jenkins.support.slowrequest;
+
+import com.cloudbees.jenkins.support.timer.FileListCap;
+import hudson.ExtensionList;
+import hudson.ExtensionPoint;
+
+import java.io.IOException;
+import java.util.List;
+
+public abstract class SlowRequest implements ExtensionPoint {
```
Comment (oleg-nenashev) [inline src/main/java/com/cloudbees/jenkins/support/slowrequest/SlowRequest.java]: Needs to be documented 🐛

File: src/main/java/com/cloudbees/jenkins/support/slowrequest/SlowRequest.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    abstract void doRun(InflightRequest req, long totalTime, FileListCap logs) throws IOException;
+
+    public static List<SlowRequest> all() {
```
Comment (oleg-nenashev) [inline src/main/java/com/cloudbees/jenkins/support/slowrequest/SlowRequest.java]: Why not ExtensionList?

File: src/main/java/com/cloudbees/jenkins/support/slowrequest/SlowRequest.java (hunk 1 above)
Comment (oleg-nenashev) [inline src/main/java/com/cloudbees/jenkins/support/slowrequest/SlowRequest.java]: I also doubt the method is ideal. Maybe it should be `AsyncMonitoringRequest` or whatever

File: src/main/java/com/cloudbees/jenkins/support/slowrequest/ThreadDumpSlowRequest.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+@Extension
+public class ThreadDumpSlowRequest extends SlowRequest {
+
+    private final SimpleDateFormat format = new SimpleDateFormat("yyyyMMdd-HHmmss.SSS");
+    {
+        format.setTimeZone(TimeZone.getTimeZone("UTC"));
+    }
+
+    @Override
+    public void doRun(InflightRequest req, long totalTime, FileListCap logs) throws IOException {
```
Comment (oleg-nenashev) [inline src/main/java/com/cloudbees/jenkins/support/slowrequest/ThreadDumpSlowRequest.java]: I am not sure what is purpose of moving `totalTime` in such way

**Comment by jglick** [comment]
Definitely needs a downstream PR showing this being used for something.

**Comment by christ66** [comment]
@jglick It is currently being used for `ThreadDumpSlowRequest` class in this pull request. Should I provide a second use case for the extension point?

**Pull request closed without merging**